Frameworks/Foundation/NSDecimal.mm

```diff
@@ -0,0 +1,592 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//******************************************************************************
+
+#import <Foundation/NSDecimal.h>
+#import <StubReturn.h>
+#include <algorithm>
+
+#define EXPONENT_MAX 127
+#define EXPONENT_MIN -128
+
+#define MANTISSA(decimal) decimal->_mantissa
+
+#define LENGTH(decimal) decimal->_length
+
+#define EXPONENT(decimal) decimal->_exponent
+
+#define NEGATIVE_FLAG(decimal) decimal->_isNegative
+
+#define SET_TO_ZERO(decimal) \
+    LENGTH(decimal) = 0;     \
+    EXPONENT(decimal) = 0;   \
+    NEGATIVE_FLAG(decimal) = 0
+
+#define IS_NEGATIVE(decimal) decimal->_isNegative
+
+#define IS_POSITIVE(decimal) !IS_NEGATIVE(decimal)
+
+#define IS_ZERO(decimal) (decimal->_length == 0) && IS_POSITIVE(decimal)
+
+#define INVALIDATE_DECIMAL(decimal) decimal->_reserved = 1
+
+#define VALIDATE_ARITHMETIC_INPUT(result, leftOperand, rightOperand)                \
+    if (NSDecimalIsNotANumber(leftOperand) || NSDecimalIsNotANumber(leftOperand)) { \
+        if (result) {                                                               \
+            INVALIDATE_DECIMAL(result);                                             \
+        }                                                                           \
+        return NSCalculationNoError;                                                \
+    }
+
+#define CHECK_DIVIDE_BY0(decimal)                           \
+    NSDecimal zero = { 0 };                                 \
+    if (NSDecimalCompare(decimal, zero) == NSOrderedSame) { \
+        return NSCalculationDivideByZero;                   \
+    }
+
+/**
+ @Status Interoperable
+*/
+BOOL NSDecimalIsNotANumber(const NSDecimal* dcm) {
+    THROW_NS_IF_FALSE(E_INVALIDARG, dcm != NULL);
+    return (dcm->_reserved == 1);
+}
+
+/**
+ @Status Stub
+ @Notes
+*/
+NSString* NSDecimalString(const NSDecimal* dcm, id locale) {
+    UNIMPLEMENTED();
+    return StubReturn();
+}
+
+static NSCalculationError _addSameSign(NSDecimal* result, NSDecimal* x, NSDecimal* y) {
+    // add starting at the end of the larger number.
+    // Note: we will have different placements of digits, since the decimal would of been compacted.
+    // We could ideally do a placement shifting to make the indexing much less complicated, but we would end up doing extra
+    // mantissa normalization and then just end up recompacting it.
+    NSDecimal* shortest;
+    if (NSDecimalCompare(x, y) == NSOrderedDescending) {
+        NSDecimalCopy(result, x);
+        shortest = y;
+    } else {
+        NSDecimalCopy(result, y);
+        shortest = x;
+    }
+
+    int carryOver = 0;
+    int offset = (LENGTH(result) - LENGTH(shortest));
+    for (int shortIndex = LENGTH(shortest) - 1; shortIndex >= 0; --shortIndex) {
```

Comment:
it could be 0, in that case you will not enter the loop and exit with the result as y.

---
In reply to: [75562017](https://github.com/Microsoft/WinObjC/pull/896#discussion_r75562017) [](ancestors = 75562017)